3.x: Mypy Unreachable

The unreachable check was disabled in https://github.com/marshmallow-code/marshmallow/pull/2878 after https://github.com/marshmallow-code/marshmallow/actions/runs/20423614307/job/58679701675 blocked the pypi 3.36.2 security release.

> src/marshmallow/fields.py:627: error: Statement is unreachable  [unreachable]

When I remove the offending type cast it reveals the original error the cast silenced:

> src/marshmallow/fields.py:534: error: Incompatible types in assignment (expression has type "Schema | str | dict[str, Field[Any]] | type", variable has type "Schema | SchemaMeta | dict[str, Field[Any]]")  [assignment]

The argument hint is missing `str` in the callable return type. The code correctly finds the class from a callable that returns a string, so that is an easy fix.

The other difference is that mypy turned `SchemaMeta` into `type` in the argument type at the top level, but not in the return type of the callable.`type` seems to be caused by the `not isinstance(..., type)` check shadowing `SchemaMeta` as an alias for `type`. mypy can't implicitly union the branched types here, so it assumes the type in the first branch and complains about the second assignment not being compatible with it. Possibly a mypy bug. The proper fix is to `nested = self.nested` to establish the broadest type, and the callable check safely narrows it.

`not isinstance(..., type)` can then be switched to `not isinstance(..., SchemaMeta)` to clarify it is deferring schema construction.

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -8,6 +8,7 @@ Bug fixes:
 
 - `marshmallow.validate.URL` accepts Internationalized Domain Names (IDNs) (:issue:`2821`).
   Thanks :user:`touhidurrr` for the report.
+- Typing: Fix typing of ``netsted`` in `marshmallow.fields.Nested` (:pr:`2935`).
 
 4.2.3 (2026-03-25)
 ------------------
@@ -26,7 +27,7 @@ Bug fixes:
   Thanks :user:`bysiber` for the PR.
 - `marshmallow.fields.DateTime` with ``format="timestamp_ms"`` properly
   rejects bool values (:pr:`2904`). Thanks :user:`bysiber` for the PR.
-- Fix typing of ``error_essages`` argument to `marshmallow.fields.Field` (:pr:`1636`).
+- Fix typing of ``error_messages`` argument to `marshmallow.fields.Field` (:pr:`1636`).
   Thanks :user:`repole` for reporting and :user:`dhruvildarji` for the PR.
 
 Other changes:

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -133,7 +133,7 @@ ignore = [
 [tool.mypy]
 files = ["src", "tests", "examples"]
 ignore_missing_imports = true
-warn_unreachable = false
+warn_unreachable = true
 warn_unused_ignores = true
 warn_redundant_casts = true
 no_implicit_optional = true

--- a/src/marshmallow/fields.py
+++ b/src/marshmallow/fields.py
@@ -500,7 +500,7 @@ class Nested(Field):
             | SchemaMeta
             | str
             | dict[str, Field]
-            | typing.Callable[[], Schema | SchemaMeta | dict[str, Field]]
+            | typing.Callable[[], Schema | SchemaMeta | str | dict[str, Field]]
         ),
         *,
         only: types.StrSequenceOrSet | None = None,

--- a/src/marshmallow/fields.py
+++ b/src/marshmallow/fields.py
@@ -529,10 +529,10 @@ class Nested(Field):
         """The nested Schema object."""
         if not self._schema:
             # defer the import of `marshmallow.schema` to avoid circular imports
-            from marshmallow.schema import Schema  # noqa: PLC0415
+            from marshmallow.schema import Schema, SchemaMeta  # noqa: PLC0415
 
             nested = self.nested
-            if callable(nested) and not isinstance(nested, type):
+            if callable(nested) and not isinstance(nested, SchemaMeta):
                 nested = nested()
 
             if isinstance(nested, dict):

--- a/src/marshmallow/fields.py
+++ b/src/marshmallow/fields.py
@@ -528,12 +528,12 @@ class Nested(Field):
     def schema(self) -> Schema:
         """The nested Schema object."""
         if not self._schema:
-            if callable(self.nested) and not isinstance(self.nested, type):
-                nested = self.nested()
-            else:
-                nested = typing.cast("Schema", self.nested)
             # defer the import of `marshmallow.schema` to avoid circular imports
             from marshmallow.schema import Schema  # noqa: PLC0415
+
+            nested = self.nested
+            if callable(nested) and not isinstance(nested, type):
+                nested = nested()
 
             if isinstance(nested, dict):
                 nested = Schema.from_dict(nested)

